### PR TITLE
Add a GitHub project badge for Hacktoberfest [GithubHacktoberfest]

### PR DESCRIPTION
**Pull request by paulmelnikow**
Something I've wanted to offer is a nice badge for Hacktoberfest, which projects can use to advertise their participation, encourage contributions, and celebrate new contributions. There was some discussion about this at https://github.com/badges/shields/issues/1519#issuecomment-456267637 and a little bit of chat in Discord today.

Here's a proposal! Would be really curious for comments!

By default this counts open issues labeled **hacktoberfest** but you can pick a different label if you want. By default it counts all PRs merged in October but you can filter that with a label if you want.

This is hard-coded for Hacktoberfest 2019. At some point if people like the badge, we'll want to generalize it, though I think we could wait and do that sometime next year.

Alternatively we could include 2019 in the badge URL, which would make it clear that projects are participating in a given year (as opposed to some past year) and also to display stats from previous years.

### CODE REVIEW

**Comment by paulmelnikow** [comment]
As I think about this more, I'm thinking the route should be `/github/hacktoberfest/2019/:user/:repo`. That way when you see the badge you know it's fresh, and the badges can be used later to show stats from the previous event.

File: services/github/github-hacktoberfest.service.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  documentation: githubDocumentation,
+  transformErrors,
+} = require('./github-helpers')
+
+const documentation = `
+  <p>
+    This badge is designed for projects hosted on GitHub which are
+    participating in
+    <a href="https://hacktoberfest.digitalocean.com">Hacktoberfest</a>,
+    an initiative to encourage participatin in open-source projects. The
```
Comment (MattIPv4) [inline services/github/github-hacktoberfest.service.js]:
```suggestion
    an initiative to encourage participating in open-source projects. The
```

File: services/github/github-hacktoberfest.service.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    badge can be added to the project readme to encourage potential
+    contributors to to review the suggested issues and to celebrate the
```
Comment (MattIPv4) [inline services/github/github-hacktoberfest.service.js]:
```suggestion
    contributors to review the suggested issues and to celebrate the
```

File: services/github/github-hacktoberfest.service.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        .join(', ') || 'is done!'
+
+    return { message }
+  }
+
+  async fetch({ user, repo, suggestionLabel = 'hacktoberfest' }) {
+    const isValidOctoberPR = [
+      `repo:${user}/${repo}`,
+      'is:pr',
+      'created:>2019-10-01',
```
Comment (MattIPv4) [inline services/github/github-hacktoberfest.service.js]: Do you want to also add a cap for 1st Nov? (It looks like the days left was capped)

**Comment by paulmelnikow** [comment]
Thanks for the comments @MattIPv4!

**Comment by calebcartwright** [comment]
I love this idea!

File: services/github/github-hacktoberfest.service.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+const documentation = `
+  <p>
+    This badge is designed for projects hosted on GitHub which are
+    participating in
+    <a href="https://hacktoberfest.digitalocean.com">Hacktoberfest</a>,
+    an initiative to encourage participating in open-source projects. The
+    badge can be added to the project readme to encourage potential
+    contributors to review the suggested issues and to celebrate the
+    contributions that have already landed.
```
Comment (MattIPv4) [inline services/github/github-hacktoberfest.service.js]:
```suggestion
    contributions that have already been submitted.
```

Landed sounds wrong as this isn't just considering merged PRs.

File: services/github/github-hacktoberfest.service.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+      {
+        title:
+          'GitHub Hacktoberfest combined status (suggestion label override)',
+        namedParams: {
+          user: 'tmrowco',
+          repo: 'tmrowapp-contrib',
+        },
+        queryParams: {
+          suggestion_label: 'help wanted',
+          contribution_label: undefined,
```
Comment (calebcartwright) [inline services/github/github-hacktoberfest.service.js]: Is this needed?

File: services/github/github-hacktoberfest.service.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+  static get defaultBadgeData() {
+    return {
+      label: 'hacktoberfest',
+      color: 'orange',
+    }
+  }
+
+  static render({ suggestedIssueCount, contributionCount, daysLeft }) {
+    if (daysLeft === undefined) {
+      daysLeft = moment('2019-11-01').diff(moment(), 'days')
```
Comment (calebcartwright) [inline services/github/github-hacktoberfest.service.js]: I suppose the days remaining count will be based on the either the time zone in France or Canada where our badge servers are deployed. I think that's fine (trying to do time zone localization seems more trouble than it'd be worth), though perhaps it'd be worth a mention in the docs?

File: services/github/github-hacktoberfest.service.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    const message =
+      [
+        suggestedIssueCount ? `${metric(suggestedIssueCount)} open issues` : '',
+        contributionCount ? `${metric(contributionCount)} PRs` : '',
+        daysLeft > 0 ? `${daysLeft} day${daysLeft > 1 ? 's' : ''} left` : '',
+      ]
+        .filter(Boolean)
+        .join(', ') || 'is done!'
```
Comment (calebcartwright) [inline services/github/github-hacktoberfest.service.js]: Nice!

File: services/github/github-hacktoberfest.service.js (hunk 6 above)
Comment (MattIPv4) [reply services/github/github-hacktoberfest.service.js]: Hacktoberfest runs from `30 SEP 2019 10:00:00 UTC` till `01 NOV 2019 12:00:00 UTC` to allow for all timezones, so using that might work best?

File: services/github/github-hacktoberfest.service.js (hunk 4 above)
Comment (paulmelnikow) [reply services/github/github-hacktoberfest.service.js]: Right-o. Gonna go with 'made' which I like slightly better.

File: services/github/github-hacktoberfest.service.js (hunk 5 above)
Comment (paulmelnikow) [reply services/github/github-hacktoberfest.service.js]: Whoops, thanks.

File: services/github/github-hacktoberfest.service.js (hunk 6 above)
Comment (paulmelnikow) [reply services/github/github-hacktoberfest.service.js]: Hmm I was thinking it'd be UTC, though I suppose you're right that it's going to be the time on the server.

I'd suggest we dodge the issue by picking a UTC end time slightly into 11/1 when we expect it's rolled over to 11/1 for most programmers in the world.

@MattIPv4 am I right that it ends at midnight in the contributor's local time on 10/31? Or is there a coordinated end time?

Comment (MattIPv4) [reply services/github/github-hacktoberfest.service.js]: Global cut-off will be `01 NOV 2019 12:00:00 UTC`.

File: services/github/github-hacktoberfest.service.js [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  }
+
+  static render({ suggestedIssueCount, contributionCount, daysLeft }) {
+    if (daysLeft === undefined) {
+      daysLeft = moment('2019-11-01').diff(moment(), 'days')
+    }
+
+    const message =
+      [
+        suggestedIssueCount ? `${metric(suggestedIssueCount)} open issues` : '',
```
Comment (calebcartwright) [inline services/github/github-hacktoberfest.service.js]:
```suggestion
        suggestedIssueCount ? `${metric(suggestedIssueCount)} open issue${suggestedIssueCount > 1 ? 's' : ''}` : '',
```

File: services/github/github-hacktoberfest.service.js (hunk 6 above)
Comment (paulmelnikow) [reply services/github/github-hacktoberfest.service.js]: Noon UTC on 11/1, sounds good.

File: services/github/github-hacktoberfest.tester.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+'use strict'
+
+const Joi = require('@hapi/joi')
+const t = (module.exports = require('../tester').createServiceTester())
+
+const isHactoberfestCombinedStatus = Joi.string().regex(
+  /^[0-9]+ open issues(, [0-9]+ PRs)?(, [0-9]+ days left)?$/
```
Comment (calebcartwright) [inline services/github/github-hacktoberfest.tester.js]: Haven't actually tested this, but just a tweak to handle the singular form of `issues` and/or `days`
```suggestion
  /^[0-9]+ open issues?(, [0-9]+ PRs)?(, [0-9]+ days? left)?$/
```

Comment (MattIPv4) [reply services/github/github-hacktoberfest.tester.js]: Doesn't `PRs` also need the switch on it?

File: services/github/github-hacktoberfest.service.js (hunk 8 above)
Comment (MattIPv4) [reply services/github/github-hacktoberfest.service.js]: Same can be applied to PRs?

File: services/github/github-hacktoberfest.tester.js (hunk 9 above)
Comment (paulmelnikow) [reply services/github/github-hacktoberfest.tester.js]: Updated, thanks!

**Pull request merged**